File: client/lib/reg/handler_registry.dart
/// Client-side handler registry. Mirrors the server's `(endpoint, action)`
/// keying scheme so the dispatcher can encode the typed `Input` to JSON and
/// decode `Output` JSON back into a typed instance.
///
/// In v1 the registry is purely a JSON codec table; it does not perform
/// authorization or any other server logic.
library;

/// Encodes a typed input into the JSON tree the server expects in `data`.
/// `null` means "no input payload" (the dispatcher omits `data` from the
/// sub-request).
typedef InputEncoder<I> = dynamic Function(I input);

/// Decodes a sub-response's `data` (already json-decoded) into a typed
/// output instance.
typedef OutputDecoder<R> = R Function(dynamic raw);

/// Type-erased codecs the dispatcher actually invokes. The registry wraps
/// the typed [InputEncoder] / [OutputDecoder] in closures so the dispatcher
/// can hold them without leaking generic type parameters all the way out.
typedef RawEncode = dynamic Function(dynamic input);
typedef RawDecode = dynamic Function(dynamic raw);

/// One registry row, typed at registration time.
class HandlerSpec<I, R> {
  final String endpoint;
  final String action;
  final InputEncoder<I> encode;
  final OutputDecoder<R> decode;

  const HandlerSpec({
    required this.endpoint,
    required this.action,
    required this.encode,
    required this.decode,
  });

  String get key => '$endpoint.$action';
}

/// Type-erased view onto a [HandlerSpec]. Created by the registry at
/// registration time; consumed by the dispatcher.
class HandlerEntry {
  final String endpoint;
  final String action;
  final RawEncode encode;
  final RawDecode decode;
  const HandlerEntry({
    required this.endpoint,
    required this.action,
    required this.encode,
    required this.decode,
  });
}

/// Process-wide registry. Phase 12 uses it as a typed lookup table; later
/// phases plug cache invalidation hints in here too.
class HandlerRegistry {
  final Map<String, HandlerEntry> _by = {};

  void register<I, R>(HandlerSpec<I, R> spec) {
    final k = spec.key;
    if (_by.containsKey(k)) {
      throw StateError('handler $k already registered');
    }
    // Wrap the typed codecs in `dynamic`-shaped closures so the dispatcher
    // can call them without any generic-variance gymnastics.
    _by[k] = HandlerEntry(
      endpoint: spec.endpoint,
      action: spec.action,
      encode: (dynamic input) => spec.encode(input as I),
      decode: (dynamic raw) => spec.decode(raw),
    );
  }

  HandlerEntry? lookup(String endpoint, String action) =>
      _by['$endpoint.$action'];

  /// True when a handler is registered for `(endpoint, action)`.
  bool has(String endpoint, String action) =>
      _by.containsKey('$endpoint.$action');
}

File: client/src/reg/handler_registry.ts
/**
 * Client-side handler registry. Mirrors the server's `(endpoint, action)`
 * keying scheme so the dispatcher can encode the typed `Input` to JSON and
 * decode `Output` JSON back into a typed instance.
 *
 * In v1 the registry is purely a JSON codec table; it does not perform
 * authorization or any other server logic. Ported from the Dart
 * `client/lib/reg/handler_registry.dart`.
 */

/**
 * Encodes a typed input into the JSON tree the server expects in `data`.
 * The encoder is expected to return a JSON-encodable value (object, array,
 * primitive). The dispatcher serialises the wrapping {@link SubRequest} for
 * the wire.
 */
export type InputEncoder<I> = (input: I) => unknown;

/**
 * Decodes a sub-response's `data` (already JSON-decoded into a JS value)
 * into a typed output instance.
 */
export type OutputDecoder<R> = (raw: unknown) => R;

/** One registry row, typed at registration time. */
export interface HandlerSpec<I, R> {
  readonly endpoint: string;
  readonly action: string;
  readonly encode: InputEncoder<I>;
  readonly decode: OutputDecoder<R>;
}

/**
 * Process-wide registry. The dispatcher looks specs up by
 * `${endpoint}.${action}` and invokes their encode/decode callbacks.
 */
export class HandlerRegistry {
  private readonly by = new Map<string, HandlerSpec<unknown, unknown>>();

  /** Register a single handler. Throws if `(endpoint, action)` already has one. */
  register<I, R>(spec: HandlerSpec<I, R>): void {
    const k = `${spec.endpoint}.${spec.action}`;
    if (this.by.has(k)) {
      throw new Error(`handler ${k} already registered`);
    }
    // Type-erase the spec so we can hold heterogeneous handlers in one map;
    // the typed-out version is reconstituted at lookup time.
    this.by.set(k, spec as unknown as HandlerSpec<unknown, unknown>);
  }

  /** Look up a handler by `(endpoint, action)`. Returns undefined when missing. */
  lookup<I, R>(endpoint: string, action: string): HandlerSpec<I, R> | undefined {
    const got = this.by.get(`${endpoint}.${action}`);
    return got as HandlerSpec<I, R> | undefined;
  }

  /** True when a handler is registered for `(endpoint, action)`. */
  has(endpoint: string, action: string): boolean {
    return this.by.has(`${endpoint}.${action}`);
  }
}
